grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+  local ntType::TypeExpr =
+    nominalTypeExpr(qNameTypeId(name, location=this.location),
+      location=this.location);
+
+  -- PARTIAL: head() and tail() are safe because names are nonempty.
+  local nameChars::[String] = explode("", name.lexeme);
+  local prodName::IdLower_t = terminal(IdLower_t,
+    implode("", toLower(head(nameChars)) :: tail(nameChars)),
+    name.location);
+
+  local prodLHS::ProductionLHS = productionLHS(
+    nameIdLower(terminal(IdLower_t, "this"), location=this.location), '::',
+    ntType, location=this.location);
+  local prodSig::ProductionSignature =
+    productionSignatureNoCL(prodLHS, '::=',
+      productionRHSNil(location=this.location), location=this.location);
+  local prodDcl::AGDcl =
+    productionDcl('abstract', 'production',
+      nameIdLower(prodName, location=this.location), prodSig, 
+      productionBody('{', productionStmtsNil(location=this.location), '}',
+        location=this.location),
+      location=this.location);
+
+  forwards to appendAGDcl(ntDcl,
+    appendAGDcl(prodDcl, fields.recordFieldAGDcl, location=this.location),
+    location=this.location);
+}
+```
+
+### `without declaring` Nonterminal Modifier
+
+Note that to support multiple records with the same field names, we don't always declare a new annotation for the field.
+The `without declaring` nonterminal modifier requests that an annotation not be declared.
+It creates an entry in the `withoutDeclaring` synthesized attribute for the annotation.
+
+```silver
+synthesized attribute withoutDeclaring::[String] occurs on NonterminalModifier,
+  NonterminalModifiers, NonterminalModifierList;
+
+terminal Without_kwd   'without'   lexer classes {KEYWORD};
+terminal Declaring_kwd 'declaring' lexer classes {KEYWORD};
+
+concrete production withoutDeclaringOne
+this::NonterminalModifier ::= 'without'  'declaring'  fieldName::IdLower_t
+{
+  propagate nonterminalModifiers;
+
+  this.errors := [];
+  this.unparse = s"without declaring ${fieldName.lexeme}";
+
+  this.withoutDeclaring = [fieldName.lexeme];
+}
+
+concrete production withoutDeclaringMany
+this::NonterminalModifier ::= 'without'  'declaring'  '{' 
+  fieldNames::FieldNames0  '}'
+{
+  propagate nonterminalModifiers;
+
+  this.errors := [];
+  this.unparse = s"without declaring {${fieldNames.unparse}}";
+
+  this.withoutDeclaring = fieldNames.names;
+}
+
+aspect default production
+this::NonterminalModifier ::=
+{ this.withoutDeclaring = []; }
+```
+
+We need the propagates for `nonterminalModifiers` because it's interfering...
+
+```silver imports
+import silver:compiler:definition:concrete_syntax only nonterminalModifiers;
+```
+
+We want to use our own `FieldNames0` nonterminal, because none of the built-in ones seem to be "just a plain list of `IdLower_t`s."
+In particular, `QNames` is a list of type-qualified `QName`s, not just plain names.
+
+```silver
+synthesized attribute names::[String];
+
+nonterminal FieldNames0 with location, unparse, names;
```

Comment:
Why is this factored into two seperate nonterminals?

Reply:
To have a left-recursive grammar with an optional trailing comma; I can add a comment about this?